File: Strivers-A2Z-DSA-Sheet/05.Recursion/3.Try Out All Combos/python/01.Palindrome partitioning.py
"""
=============================================================================
  QUESTION: 131. Palindrome Partitioning (LeetCode)
=============================================================================

  Partition s such that every substring is a palindrome. Return all partitions.
  Example: "aab" → [["a","a","b"],["aa","b"]]

  APPROACH: Backtracking — try all possible cuts.
  NOT pick/not-pick. Try ALL cut positions, only proceed if palindrome.

  DRY RUN "aab":
    s[0..0]="a" palindrome → s[1..1]="a" → s[2..2]="b" → ["a","a","b"] ✓
    s[0..1]="aa" palindrome → s[2..2]="b" → ["aa","b"] ✓
    s[0..2]="aab" NOT palindrome → skip

  Time: O(N × 2^N), Space: O(N)
=============================================================================
"""


def is_palindrome(s):
    return s == s[::-1]


def solve(i, s, temp, ans):
    if i >= len(s):
        ans.append(temp[:])
        return

    for p in range(i, len(s)):
        sub = s[i:p + 1]
        if is_palindrome(sub):
            temp.append(sub)
            solve(p + 1, s, temp, ans)
            temp.pop()


def partition(s):
    ans = []
    solve(0, s, [], ans)
    return ans


# Driver Code
if __name__ == "__main__":
    print(partition("aab"))  # [["a","a","b"],["aa","b"]]

File: Strivers-A2Z-DSA-Sheet/05.Recursion/3.Try Out All Combos/javascript/01.Palindrome partitioning.js
/*
=============================================================================
  QUESTION: 131. Palindrome Partitioning (LeetCode)
=============================================================================

  Partition string s such that every substring is a palindrome.
  Return all possible palindrome partitionings.

  Example: "aab" → [["a","a","b"], ["aa","b"]]

=============================================================================
  APPROACH: Backtracking — Try All Cuts — O(N × 2^N) Time, O(N) Space
=============================================================================

  NOT pick/not-pick. At index i, try cutting s[i..p] for every p from i to end.
  If s[i..p] is palindrome → add to partition, recurse from p+1, backtrack.

  Dry Run "aab":
    solve(0, [])
      "a" ✓ → solve(1, ["a"]) → "a" ✓ → solve(2, ["a","a"])
                                            → "b" ✓ → push ["a","a","b"] ✓
                                 "ab" ✗
      "aa" ✓ → solve(2, ["aa"]) → "b" ✓ → push ["aa","b"] ✓
      "aab" ✗
    Result: [["a","a","b"], ["aa","b"]] ✓

=============================================================================
*/

function partition(s) {
    let result = [];

    function isPalin(str) {
        let i = 0, j = str.length - 1;
        while (i < j) { if (str[i++] !== str[j--]) return false; }
        return true;
    }

    function solve(i, temp) {
        if (i >= s.length) { result.push([...temp]); return; }

        for (let p = i; p < s.length; p++) {
            let sub = s.substring(i, p + 1);
            if (isPalin(sub)) {
                temp.push(sub);
                solve(p + 1, temp);
                temp.pop();
            }
        }
    }

    solve(0, []);
    return result;
}

// ==========================================================================
// DRIVER CODE
// ==========================================================================
console.log(partition("aab"));   // [["a","a","b"],["aa","b"]]
console.log(partition("a"));     // [["a"]]
console.log(partition("aba"));   // [["a","b","a"],["aba"]]
